tests/core/test_context.py

```diff
@@ -167,3 +170,27 @@ def test_add_non_plan(empty_context: BlueskyContext) -> None:
 def test_add_non_device(empty_context: BlueskyContext) -> None:
     with pytest.raises(TypeError):
         empty_context.device("not a device")  # type: ignore
+
+
+def test_function_spec(empty_context: BlueskyContext) -> None:
+    spec = context._type_spec_for_function(has_some_params, empty_context)
+    assert spec == {"foo": ("int", 42), "bar": ("str", "bar")}
+
+
+def test_basic_type_conversion(empty_context: BlueskyContext) -> None:
+    assert context._convert_type(int, empty_context) == int
+    assert context._convert_type(dict[str, int], empty_context) == dict[str, int]
+
+
+def test_device_reference_cache(empty_context: BlueskyContext) -> None:
+    assert empty_context._reference(Movable) is empty_context._reference(Movable)
+    assert empty_context._reference(Movable) is not empty_context._reference(Readable)
+
+
+def test_reference_type_conversion(empty_context: BlueskyContext) -> None:
+    movable_ref: Type = empty_context._reference(Movable)
+    assert context._convert_type(Movable, empty_context) == movable_ref
+    assert (
+        context._convert_type(dict[Movable, list[tuple[int, Movable]]], empty_context)
+        == dict[movable_ref, list[tuple[int, movable_ref]]]  # type: ignore
```

Comment:
Again, not sure about types here. It says `movable_ref` can't be used as a type.

Reply:
Maybe, but it's being returned by a function that is annotated as returning `Type`